DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/optimize_inner_loop/src/optimize_inner_loop.cpp

```diff
@@ -0,0 +1,280 @@
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <array>
+#include <iomanip>
+#include <iostream>
+#include <random>
+#include <type_traits>
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+
+// allow the maximum random number value to be controlled from the command line
+#ifndef RAND_RANGE_MAX
+#define RAND_RANGE_MAX 3
+#endif
+
+//// constants
+constexpr int kNumKernels = 3;
+constexpr int kRandRangeMax = RAND_RANGE_MAX;
+constexpr double kProbSuccess = 1.0 / (double)(kRandRangeMax);
+
+// Declare the kernel class names globally to avoid name mangling.
+// Templating allows us to instantiate multiple versions of the kernel.
+template <int Version> class Producer;
+template <int Version> class Consumer;
+
+// Declare the pipe class name globally to avoid name mangling.
+// Templating allows us to instantiate multiple versions of pipes for each 
+// version of the kernel.
+template <int Version> class PipeClass;
+
+//
+// Submits the kernel, which is templated on the variables:
+//    Version             - The version ID of the kernel
+//    InElementUpperBound - The upperbound (inclusive) on the elements of the
+//                          'in' vector (a negative value implies no bound). In
+//                          other words: if InElementUpperBound >= 0, then
+//                          in[i] <= InElementUpperBound, for all elements
+//                          of 'in'
+//    SpecIters           - The number of speculated iterations to set for the
+//                          inner loop
+//
+template <int Version, int InElementUpperBound, int SpecIters>
+void SubmitKernels(const device_selector &selector, std::vector<int> &in,
+                   int &res, double &kernel_time_ms) {
+  // static asserts: these cause the compiler to fail if the conditions fail
+  static_assert(Version >= 0, "Invalid kernel version");
+  static_assert(SpecIters >= 0, "SpecIters must be positive");
+
+  // the pipe
+  using Pipe = pipe<PipeClass<Version>, bool>;
+
+  kernel_time_ms = 0.0;
+  int size = in.size();
+
+  try {
+    // create the device queue with profiling enabled
+    auto prop_list = property_list{ property::queue::enable_profiling() };
+    queue q(selector, dpc_common::exception_handler, prop_list);
+
+    // The input data buffer
+    buffer in_buf(in);
+
+    // The output data buffer
+    // Scalar inputs are passed to the kernel using the lambda capture,
+    // but a SYCL buffer must be used to return a scalar from the kernel.
+    buffer<int, 1> res_buf(&res, 1);
+
+    // submit the Producer kernel
+    event p_e = q.submit([&](handler &h) {
+      // the input buffer accessor
+      accessor in_accessor{ in_buf, h, read_only };
```

Comment:
We want the kernel to be simplest.   So the accessor should be named "in".   This would make the original array named "in_data" or similar.

Reply:
went with `in_a`